File: python/AL-assignment2.py
from typing import List

def abs(n):
    if n < 0:
        return -n
    else:
        return n

def cal(arr, d):
    _sum = 0
    for i in arr:
        _sum += abs(i - d)
    return _sum

def quick_sort(arr):
    if len(arr) <= 1:
        return arr
    target = arr[len(arr) // 2]
    less, eq, large = [], [], []

    # partition
    for num in arr:
        if num < target:
            less.append(num)
        elif num > target:
            large.append(num)
        else:
            eq.append(num)

    # sum
    return quick_sort(less) + eq + quick_sort(large)

def fun(arr : List[int], n : int):
    if n == 0:
        return 0
    if n == 1:
        return arr[0]
        
    arr = quick_sort(arr)
    ans = cal(arr, arr[n // 2])
    return ans

if __name__ == "__main__":
    pass
    # fun(test, len(test))

File: C/AL-assignment2.c
#include <stdio.h>

void swap(int *arr, int i, int j) {
    int temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
}

void quickSort(int *arr, int start, int end) {
    if (start >= end) return;
    int target = arr[end];
    int i = start - 1;
    for (int j = start; j < end; j++) {
        if (arr[j] <= target) {
            swap(arr, ++i, j);
        }
    }
    swap(arr, i + 1, end);
    quickSort(arr, i + 1, end);
    quickSort(arr, start, i);
}

int calculate(int *arr, int len) {
    quickSort(arr, 0, len - 1);
    int res = 0;
    int k = arr[len / 2];
    for (int i = 0; i < len; ++i) {
        int a = arr[i] - k;
        if (a > 0)
            res += a;
        else
            res -= a;
    }
    return res;
}

int main() {
    int len = 8;
    int data[8] = {2, 1, 3, 123, 1, 99, 1, 2};
    printf("%d\n", calculate(data, len));
}